crates/torii/libp2p/src/client/mod.rs

```diff
@@ -0,0 +1,217 @@
+        let (command_sender, command_receiver) = futures::channel::mpsc::unbounded();
+        Ok(Self {
+            command_sender,
+            message_receiver,
+            event_loop: EventLoop { swarm, message_sender, command_receiver },
+        })
+    }
+}
+
+impl EventLoop {
+    pub async fn run(&mut self) {
+        loop {
+            // Poll the swarm for new events.
+            select! {
+                command = self.command_receiver.select_next_some() => {
+                    match command {
+                        Command::Subscribe(room) => {
+                            self.subscribe(&room).expect("Failed to subscribe");
+                        },
+                        Command::Unsubscribe(room) => {
+                            self.unsubscribe(&room).expect("Failed to unsubscribe");
+                        },
+                        Command::Publish(message) => {
+                            self.publish(&message).expect("Failed to publish message");
+                        },
+                    }
+                },
+                event = self.swarm.select_next_some() => {
+                    match event {
+                        SwarmEvent::Behaviour(event) => {
+                            // Handle behaviour events.
+                            if let ClientEvent::Gossipsub(gossipsub::Event::Message {
+                                propagation_source: peer_id,
+                                message_id,
+                                message,
+                            }) = event
+                            {
+                                // deserialize message
+                                let message: ServerMessage = serde_json::from_slice(&message.data)
+                                    .expect("Failed to deserialize message");
+                                self.message_sender.send((peer_id, message_id, message)).await.expect("Failed to send message");
+                            }
+                        }
+                        SwarmEvent::ConnectionClosed { cause: Some(cause), .. } => {
+                            info!(target: "torii::relay::client", cause = ?cause, "Connection closed");
+
+                            if let libp2p::swarm::ConnectionError::KeepAliveTimeout = cause {
+                                info!(target: "torii::relay::client", "Connection closed due to keep alive timeout. Shutting down client.");
+                                return;
+                            }
+                        }
+                        evt => {
+                            info!(target: "torii::relay::client", event = ?evt, "Unhandled event");
+                        }
+                    }
+                },
+            }
+        }
+    }
+
+    fn subscribe(&mut self, room: &str) -> Result<bool, Error> {
+        let topic = IdentTopic::new(room);
+        let sub = self.swarm.behaviour_mut().gossipsub.subscribe(&topic)?;
+        // if sub {
```

Comment:
remove or?

Reply:
Mhh do we want to keep track of the subscribed topics? They're hashed when subscribed to so we can't get the topic names we subscribed to, only the hash.